File: app/src/main/java/com/fastaccess/provider/timeline/handler/BetterLinkMovementExtended.java
package com.fastaccess.provider.timeline.handler;

import android.content.Context;
import android.graphics.RectF;
import android.text.Layout;
import android.text.Selection;
import android.text.Spannable;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.text.style.BackgroundColorSpan;
import android.text.style.ClickableSpan;
import android.text.style.URLSpan;
import android.text.util.Linkify;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;


/**
 * Created by Kosh on 23 Apr 2017, 12:09 PM
 * <p>
 * credit to to https://github.com/Saketme/Better-Link-Movement-Method
 */

public class BetterLinkMovementExtended extends LinkMovementMethod {
    private static final Class SPAN_CLASS = ClickableSpan.class;
    private static final int LINKIFY_NONE = -2;
    private BetterLinkMovementExtended.OnLinkClickListener onLinkClickListener;
    private BetterLinkMovementExtended.OnLinkLongClickListener onLinkLongClickListener;
    private final RectF touchedLineBounds = new RectF();
    private boolean isUrlHighlighted;
    private boolean touchStartedOverLink;
    private int activeTextViewHashcode;

    private final GestureDetector gestureDetector;
    private final LinkClickGestureListener clickGestureListener = new LinkClickGestureListener();

    private BetterLinkMovementExtended(Context context) {
        gestureDetector = new GestureDetector(context, clickGestureListener);
    }

    private final class LinkClickGestureListener extends GestureDetector.SimpleOnGestureListener {
        private GestureDetector.SimpleOnGestureListener listener = null;

        @Override public boolean onDown(MotionEvent e) {
            if(listener != null) listener.onDown(e);
            return true;
        }

        @Override public boolean onSingleTapUp(MotionEvent e) {
            return listener == null || listener.onSingleTapUp(e);
        }

        @Override public void onLongPress(MotionEvent e) {
            if(listener != null) listener.onLongPress(e);
        }
    }

    private static BetterLinkMovementExtended linkify(int linkifyMask, TextView textView) {
        BetterLinkMovementExtended movementMethod = new BetterLinkMovementExtended(textView.getContext());
        addLinks(linkifyMask, movementMethod, textView);
        return movementMethod;
    }

    public static BetterLinkMovementExtended linkifyHtml(TextView textView) {
        return linkify(LINKIFY_NONE, textView);
    }

    private static BetterLinkMovementExtended linkify(int linkifyMask, ViewGroup viewGroup) {
        BetterLinkMovementExtended movementMethod = new BetterLinkMovementExtended(viewGroup.getContext());
        rAddLinks(linkifyMask, viewGroup, movementMethod);
        return movementMethod;
    }

    public static BetterLinkMovementExtended linkifyHtml(ViewGroup viewGroup) {
        return linkify(LINKIFY_NONE, viewGroup);
    }

    public void setOnLinkClickListener(OnLinkClickListener onLinkClickListener) {
        this.onLinkClickListener = onLinkClickListener;
    }

    public void setOnLinkLongClickListener(OnLinkLongClickListener onLinkLongClickListener) {
        this.onLinkLongClickListener = onLinkLongClickListener;
    }

    private static void rAddLinks(int linkifyMask, ViewGroup viewGroup, BetterLinkMovementExtended movementMethod) {
        for (int i = 0; i < viewGroup.getChildCount(); ++i) {
            View child = viewGroup.getChildAt(i);
            if (child instanceof ViewGroup) {
                rAddLinks(linkifyMask, (ViewGroup) child, movementMethod);
            } else if (child instanceof TextView) {
                TextView textView = (TextView) child;
                addLinks(linkifyMask, movementMethod, textView);
            }
        }

    }

    private static void addLinks(int linkifyMask, BetterLinkMovementExtended movementMethod, TextView textView) {
        textView.setMovementMethod(movementMethod);
        if (linkifyMask != LINKIFY_NONE) {
            Linkify.addLinks(textView, linkifyMask);
        }

    }

    public boolean onTouchEvent(TextView view, Spannable text, MotionEvent event) {
        if (this.activeTextViewHashcode != view.hashCode()) {
            this.activeTextViewHashcode = view.hashCode();
            view.setAutoLinkMask(0);
        }

        BetterLinkMovementExtended.ClickableSpanWithText touchedClickableSpan = this.findClickableSpanUnderTouch(view, text, event);
        if (touchedClickableSpan != null) {
            this.highlightUrl(view, touchedClickableSpan, text);
        } else {
            this.removeUrlHighlightColor(view);
        }

        clickGestureListener.listener = new GestureDetector.SimpleOnGestureListener() {
            @Override public boolean onDown(MotionEvent e) {
                touchStartedOverLink = touchedClickableSpan != null;
                return true;
            }

            @Override public boolean onSingleTapUp(MotionEvent e) {
                if (touchedClickableSpan != null && touchStartedOverLink) {
                    dispatchUrlClick(view, touchedClickableSpan);
                    removeUrlHighlightColor(view);
                }

                touchStartedOverLink = false;
                return true;
            }

            @Override public void onLongPress(MotionEvent e) {
                if (touchedClickableSpan != null && touchStartedOverLink) {
                    dispatchUrlLongClick(view, touchedClickableSpan);
                    removeUrlHighlightColor(view);
                }

                touchStartedOverLink = false;
            }
        };

        boolean ret = gestureDetector.onTouchEvent(event);

        if(!ret && event.getAction() == MotionEvent.ACTION_UP) {
            clickGestureListener.listener = null;
            removeUrlHighlightColor(view);
            this.touchStartedOverLink = false;
            ret = true;
        }

        return ret;
    }

    private BetterLinkMovementExtended.ClickableSpanWithText findClickableSpanUnderTouch(TextView textView, Spannable text, MotionEvent event) {
        int touchX = (int) event.getX();
        int touchY = (int) event.getY();
        touchX -= textView.getTotalPaddingLeft();
        touchY -= textView.getTotalPaddingTop();
        touchX += textView.getScrollX();
        touchY += textView.getScrollY();
        Layout layout = textView.getLayout();
        int touchedLine = layout.getLineForVertical(touchY);
        int touchOffset = layout.getOffsetForHorizontal(touchedLine, (float) touchX);
        this.touchedLineBounds.left = layout.getLineLeft(touchedLine);
        this.touchedLineBounds.top = (float) layout.getLineTop(touchedLine);
        this.touchedLineBounds.right = layout.getLineWidth(touchedLine) + this.touchedLineBounds.left;
        this.touchedLineBounds.bottom = (float) layout.getLineBottom(touchedLine);
        if (this.touchedLineBounds.contains((float) touchX, (float) touchY)) {
            Object[] spans = text.getSpans(touchOffset, touchOffset, SPAN_CLASS);
            for (Object span : spans) {
                if (span instanceof ClickableSpan) {
                    return ClickableSpanWithText.ofSpan(textView, (ClickableSpan) span);
                }
            }
            return null;
        } else {
            return null;
        }
    }

    private void highlightUrl(TextView textView, BetterLinkMovementExtended.ClickableSpanWithText spanWithText, Spannable text) {
        if (!this.isUrlHighlighted) {
            this.isUrlHighlighted = true;
            int spanStart = text.getSpanStart(spanWithText.span());
            int spanEnd = text.getSpanEnd(spanWithText.span());
            Selection.removeSelection(text);
            text.setSpan(new BackgroundColorSpan(textView.getHighlightColor()), spanStart, spanEnd, Spanned.SPAN_INCLUSIVE_INCLUSIVE);
            textView.setText(text);
            Selection.setSelection(text, spanStart, spanEnd);
        }
    }

    private void removeUrlHighlightColor(TextView textView) {
        if (this.isUrlHighlighted) {
            this.isUrlHighlighted = false;
            Spannable text = (Spannable) textView.getText();
            BackgroundColorSpan[] highlightSpans = text.getSpans(0, text.length(), BackgroundColorSpan.class);
            for (BackgroundColorSpan highlightSpan : highlightSpans) {
                text.removeSpan(highlightSpan);
            }
            try {
                textView.setText(text);
                Selection.removeSelection(text);
            } catch (Exception ignored) {}
        }
    }

    private void dispatchUrlClick(TextView textView, BetterLinkMovementExtended.ClickableSpanWithText spanWithText) {
        String spanUrl = spanWithText.text();
        boolean handled = this.onLinkClickListener != null && this.onLinkClickListener.onClick(textView, spanUrl);
        if (!handled) {
            spanWithText.span().onClick(textView);
        }

    }

    private void dispatchUrlLongClick(TextView textView, BetterLinkMovementExtended.ClickableSpanWithText spanWithText) {
        String spanUrl = spanWithText.text();
        if(onLinkLongClickListener != null) onLinkLongClickListener.onLongClick(textView, spanUrl);
    }

    static class ClickableSpanWithText {
        private ClickableSpan span;
        private String text;

        static BetterLinkMovementExtended.ClickableSpanWithText ofSpan(TextView textView, ClickableSpan span) {
            Spanned s = (Spanned) textView.getText();
            String text;
            if (span instanceof URLSpan) {
                text = ((URLSpan) span).getURL();
            } else {
                int start = s.getSpanStart(span);
                int end = s.getSpanEnd(span);
                text = s.subSequence(start, end).toString();
            }
            return new BetterLinkMovementExtended.ClickableSpanWithText(span, text);
        }

        private ClickableSpanWithText(ClickableSpan span, String text) {
            this.span = span;
            this.text = text;
        }

        ClickableSpan span() {
            return this.span;
        }

        String text() {
            return this.text;
        }
    }

    public interface OnLinkClickListener {
        boolean onClick(TextView view, String link);
    }

    public interface OnLinkLongClickListener {
        boolean onLongClick(TextView view, String link);
    }
}

File: markdown/src/main/java/com/fastaccess/markdown/BetterLinkMovementExtended.kt
package com.fastaccess.markdown

import android.content.Context
import android.graphics.RectF
import android.text.Selection
import android.text.Spannable
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.BackgroundColorSpan
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.text.util.Linkify
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ViewGroup
import android.widget.TextView


/**
 * Created by Kosh on 23 Apr 2017, 12:09 PM
 *
 *
 * credit to to https://github.com/Saketme/Better-Link-Movement-Method
 */

class BetterLinkMovementExtended private constructor(context: Context) : LinkMovementMethod() {
    private var onLinkClickListener: ((link: String) -> Unit)? = null
    private var onLinkLongClickListener: ((link: String) -> Unit)? = null
    private val touchedLineBounds = RectF()
    private var isUrlHighlighted: Boolean = false
    private var touchStartedOverLink: Boolean = false
    private var activeTextViewHashcode: Int = 0

    private val gestureDetector: GestureDetector
    private val clickGestureListener = LinkClickGestureListener()

    init {
        gestureDetector = GestureDetector(context, clickGestureListener)
    }

    private inner class LinkClickGestureListener : GestureDetector.SimpleOnGestureListener() {
        var listener: GestureDetector.SimpleOnGestureListener? = null

        override fun onDown(e: MotionEvent): Boolean {
            listener?.onDown(e)
            return true
        }

        override fun onSingleTapUp(e: MotionEvent): Boolean {
            return listener == null || listener?.onSingleTapUp(e) == true
        }

        override fun onLongPress(e: MotionEvent) {
            listener?.onLongPress(e)
        }
    }

    fun setOnLinkClickListener(onLinkClickListener: ((link: String) -> Unit)? = null) {
        this.onLinkClickListener = onLinkClickListener
    }

    fun setOnLinkLongClickListener(onLinkClickListener: ((link: String) -> Unit)? = null) {
        this.onLinkLongClickListener = onLinkClickListener
    }

    override fun onTouchEvent(view: TextView, text: Spannable, event: MotionEvent): Boolean {
        if (this.activeTextViewHashcode != view.hashCode()) {
            this.activeTextViewHashcode = view.hashCode()
            view.autoLinkMask = 0
        }

        val touchedClickableSpan = this.findClickableSpanUnderTouch(view, text, event)
        if (touchedClickableSpan != null) {
            this.highlightUrl(view, touchedClickableSpan, text)
        } else {
            this.removeUrlHighlightColor(view)
        }

        clickGestureListener.listener = object : GestureDetector.SimpleOnGestureListener() {
            override fun onDown(e: MotionEvent): Boolean {
                touchStartedOverLink = touchedClickableSpan != null
                return true
            }

            override fun onSingleTapUp(e: MotionEvent): Boolean {
                if (touchedClickableSpan != null && touchStartedOverLink) {
                    dispatchUrlClick(view, touchedClickableSpan)
                    removeUrlHighlightColor(view)
                }

                touchStartedOverLink = false
                return true
            }

            override fun onLongPress(e: MotionEvent) {
                if (touchedClickableSpan != null && touchStartedOverLink) {
                    dispatchUrlLongClick(view, touchedClickableSpan)
                    removeUrlHighlightColor(view)
                }

                touchStartedOverLink = false
            }
        }

        var ret = gestureDetector.onTouchEvent(event)

        if (!ret && event.action == MotionEvent.ACTION_UP) {
            clickGestureListener.listener = null
            removeUrlHighlightColor(view)
            this.touchStartedOverLink = false
            ret = true
        }

        return ret
    }

    private fun findClickableSpanUnderTouch(textView: TextView, text: Spannable,
                                            event: MotionEvent): BetterLinkMovementExtended.ClickableSpanWithText? {
        var touchX = event.x.toInt()
        var touchY = event.y.toInt()
        touchX -= textView.totalPaddingLeft
        touchY -= textView.totalPaddingTop
        touchX += textView.scrollX
        touchY += textView.scrollY
        val layout = textView.layout
        val touchedLine = layout.getLineForVertical(touchY)
        val touchOffset = layout.getOffsetForHorizontal(touchedLine, touchX.toFloat())
        this.touchedLineBounds.left = layout.getLineLeft(touchedLine)
        this.touchedLineBounds.top = layout.getLineTop(touchedLine).toFloat()
        this.touchedLineBounds.right = layout.getLineWidth(touchedLine) + this.touchedLineBounds.left
        this.touchedLineBounds.bottom = layout.getLineBottom(touchedLine).toFloat()
        if (this.touchedLineBounds.contains(touchX.toFloat(), touchY.toFloat())) {
            val spans = text.getSpans<ClickableSpan>(touchOffset, touchOffset, SPAN_CLASS)
            for (span in spans) {
                if (span is ClickableSpan) {
                    return ClickableSpanWithText.ofSpan(textView, span)
                }
            }
            return null
        } else {
            return null
        }
    }

    private fun highlightUrl(textView: TextView, spanWithText: BetterLinkMovementExtended.ClickableSpanWithText?, text: Spannable) {
        if (!this.isUrlHighlighted) {
            this.isUrlHighlighted = true
            val spanStart = text.getSpanStart(spanWithText!!.span())
            val spanEnd = text.getSpanEnd(spanWithText.span())
            Selection.removeSelection(text)
            text.setSpan(BackgroundColorSpan(textView.highlightColor), spanStart, spanEnd, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
            textView.text = text
            Selection.setSelection(text, spanStart, spanEnd)
        }
    }

    private fun removeUrlHighlightColor(textView: TextView) {
        if (this.isUrlHighlighted) {
            this.isUrlHighlighted = false
            val text = textView.text as Spannable
            val highlightSpans = text.getSpans(0, text.length, BackgroundColorSpan::class.java)
            for (highlightSpan in highlightSpans) {
                text.removeSpan(highlightSpan)
            }
            try {
                textView.text = text
                Selection.removeSelection(text)
            } catch (ignored: Exception) {
            }

        }
    }

    private fun dispatchUrlClick(textView: TextView, spanWithText: BetterLinkMovementExtended.ClickableSpanWithText) {
        val spanUrl = spanWithText.text()
        val handled = this.onLinkClickListener != null
        if (!handled) {
            spanWithText.span().onClick(textView)
        } else {
            this.onLinkClickListener?.invoke(spanUrl)
        }

    }

    private fun dispatchUrlLongClick(textView: TextView, spanWithText: BetterLinkMovementExtended.ClickableSpanWithText) {
        val spanUrl = spanWithText.text()
        if (onLinkLongClickListener != null) this.onLinkLongClickListener?.invoke(spanUrl)
    }

    internal class ClickableSpanWithText private constructor(private val span: ClickableSpan, private val text: String) {

        fun span(): ClickableSpan {
            return this.span
        }

        fun text(): String {
            return this.text
        }

        companion object {

            fun ofSpan(textView: TextView, span: ClickableSpan): BetterLinkMovementExtended.ClickableSpanWithText {
                val s = textView.text as Spanned
                val text: String
                if (span is URLSpan) {
                    text = span.url
                } else {
                    val start = s.getSpanStart(span)
                    val end = s.getSpanEnd(span)
                    text = s.subSequence(start, end).toString()
                }
                return BetterLinkMovementExtended.ClickableSpanWithText(span, text)
            }
        }
    }

    companion object {
        private val SPAN_CLASS = ClickableSpan::class.java
        private const val LINKIFY_NONE = -2

        private fun linkify(linkifyMask: Int, textView: TextView): BetterLinkMovementExtended {
            val movementMethod = BetterLinkMovementExtended(textView.context)
            addLinks(linkifyMask, movementMethod, textView)
            return movementMethod
        }

        fun linkifyHtml(textView: TextView): BetterLinkMovementExtended {
            return linkify(LINKIFY_NONE, textView)
        }

        private fun linkify(linkifyMask: Int, viewGroup: ViewGroup): BetterLinkMovementExtended {
            val movementMethod = BetterLinkMovementExtended(viewGroup.context)
            rAddLinks(linkifyMask, viewGroup, movementMethod)
            return movementMethod
        }

        fun linkifyHtml(viewGroup: ViewGroup): BetterLinkMovementExtended {
            return linkify(LINKIFY_NONE, viewGroup)
        }

        private fun rAddLinks(linkifyMask: Int, viewGroup: ViewGroup, movementMethod: BetterLinkMovementExtended) {
            for (i in 0 until viewGroup.childCount) {
                val child = viewGroup.getChildAt(i)
                if (child is ViewGroup) {
                    rAddLinks(linkifyMask, child, movementMethod)
                } else if (child is TextView) {
                    addLinks(linkifyMask, movementMethod, child)
                }
            }

        }

        private fun addLinks(linkifyMask: Int, movementMethod: BetterLinkMovementExtended, textView: TextView) {
            textView.movementMethod = movementMethod
            if (linkifyMask != LINKIFY_NONE) {
                Linkify.addLinks(textView, linkifyMask)
            }

        }
    }
}